File: zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/openvino/Predict.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.examples.vnni.openvino

import com.intel.analytics.bigdl.dataset.SampleToMiniBatch
import com.intel.analytics.bigdl.nn.abstractnn.Activity
import com.intel.analytics.bigdl.numeric.NumericFloat
import com.intel.analytics.bigdl.transform.vision.image.ImageFeature
import com.intel.analytics.zoo.common.NNContext
import com.intel.analytics.zoo.feature.image.{ImageCenterCrop, ImageMatToTensor, ImageResize, ImageSet, ImageSetToSample}
import com.intel.analytics.zoo.models.image.imageclassification.{LabelOutput, LabelReader}
import com.intel.analytics.zoo.pipeline.inference.InferenceModel
import org.apache.log4j.{Level, Logger}
import org.opencv.imgcodecs.Imgcodecs
import scopt.OptionParser

case class PredictParams(folder: String = "./",
                         model: String = "",
                         weight: String = "",
                         batchSize: Int = 4,
                         topN: Int = 5,
                         partitionNum: Int = 4,
                         isInt8: Boolean = false)

object Predict {
  Logger.getLogger("org").setLevel(Level.ERROR)
  Logger.getLogger("akka").setLevel(Level.ERROR)
  Logger.getLogger("breeze").setLevel(Level.ERROR)
  Logger.getLogger("com.intel.analytics.zoo.feature.image").setLevel(Level.ERROR)
  Logger.getLogger("com.intel.analytics.zoo").setLevel(Level.INFO)

  val logger: Logger = Logger.getLogger(getClass)

  def main(args: Array[String]): Unit = {
    val parser = new OptionParser[PredictParams]("ResNet50 Int8 Inference Example") {
      opt[String]('f', "folder")
        .text("The path to the image data")
        .action((x, c) => c.copy(folder = x))
        .required()
      opt[String]('m', "model")
        .text("The path to the int8 ResNet50 model")
        .action((x, c) => c.copy(model = x))
        .required()
      opt[String]('w', "weight")
        .text("The path to the int8 ResNet50 model weight")
        .action((x, c) => c.copy(weight = x))
      opt[Int]("topN")
        .text("top N number")
        .action((x, c) => c.copy(topN = x))
      opt[Int]("partitionNum")
        .text("The number of partitions to cut the dataset into")
        .action((x, c) => c.copy(partitionNum = x))
      opt[Int]('b', "batchSize")
        .text("batch size")
        .action((x, c) => c.copy(batchSize = x))
      opt[Boolean]("isInt8")
        .text("Is Int8 optimized model?")
        .action((x, c) => c.copy(isInt8 = x))
    }
    parser.parse(args, PredictParams()).map(param => {
      val sc = NNContext.initNNContext("Predict Image with OpenVINO Int8 Example")

      val model = new InferenceModel(1)

      if (param.isInt8) {
        model.doLoadOpenVINOInt8(param.model, param.weight, param.batchSize)
      } else {
        model.doLoadOpenVINO(param.model, param.weight)
      }

      // Read ImageNet val
      val images = ImageSet.read(param.folder,
        imageCodec = Imgcodecs.CV_LOAD_IMAGE_COLOR).toLocal()

      val start = System.nanoTime()
      // Pre-processing
      val inputs = images ->
        ImageResize(256, 256) ->
        ImageCenterCrop(224, 224) ->
        ImageMatToTensor(shareBuffer = false) ->
        ImageSetToSample()
      val batched = inputs.toDataSet() -> SampleToMiniBatch(param.batchSize)

      val preProcessTime = System.nanoTime() - start
      val imageNum = images.array.length
      val batchNum = batched.toLocal().data(false).size

      // Predict
      logger.debug("Begin Prediction")
      val predicts = batched.toLocal().data(false).flatMap(miniBatch => {
        val predict = model.doPredict(miniBatch
          .getInput.toTensor.addSingletonDimension())
        predict.toTensor.squeeze.split(1).asInstanceOf[Array[Activity]]
      })
      // Add prediction into imageset
      images.array.zip(predicts.toIterable).foreach(tuple => {
        tuple._1(ImageFeature.predict) = tuple._2
      })
      // Transform prediction into Labels and probs
      val labelOutput = LabelOutput(LabelReader.apply("IMAGENET"))
      val results = labelOutput(images).toLocal().array

      // Output results
      results.foreach(imageFeature => {
        logger.info(s"image: ${imageFeature.uri}, top ${param.topN}")
        val classes = imageFeature("classes").asInstanceOf[Array[String]]
        val probs = imageFeature("probs").asInstanceOf[Array[Float]]
        for (i <- 0 until param.topN) {
          logger.info(s"\t class: ${classes(i)}, credit: ${probs(i)}")
        }
      })
      logger.info(s"Prediction finished.")

      val timeUsed = System.nanoTime() - start
      // Post-processing
      val throughput = "%.2f"
        .format(imageNum / (timeUsed / 1e9))
      val predictThroughput = "%.2f"
        .format(imageNum / ((timeUsed - preProcessTime)/ 1e9))
      val batchPredictLatency = "%.2f"
        .format((timeUsed - preProcessTime) / 1e6 / batchNum)
      val batchLatency = "%.2f".format(timeUsed / 1e6 / batchNum)
      logger.info(s"Pre-Processing: Takes ${preProcessTime / 1e6} ms")
      logger.info(s"Predict: Takes ${(timeUsed - preProcessTime) / 1e6} ms, " +
        s"throughput is ${predictThroughput} FPS (imgs/sec)")
      logger.info(s"Average Batch Predict latency is $batchPredictLatency ms")
      logger.info(s"End to End: Takes ${timeUsed / 1e6} ms, " +
        s"throughput is $throughput FPS (imgs/sec)")
      logger.info(s"Average End to End Batch Predict latency is $batchLatency ms")
      // Evaluation
      // Compare labels and output results
      sc.stop()
    })
  }
}

File: pyzoo/zoo/examples/openvino/mask_rcnn/predict.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys
import numpy as np
import random
import cv2
from os.path import join
from optparse import OptionParser

from zoo.common.nncontext import init_nncontext
from zoo.feature.image import ImageSet
from zoo.pipeline.inference import InferenceModel


def random_colour_masks(image):
    colours = [[0, 255, 0],
               [0, 0, 255], [255, 0, 0], [0, 255, 255], [255, 255, 0], [255, 0, 255], [80, 70, 180], [250, 80, 190],
               [245, 145, 50], [70, 150, 250], [50, 190, 190]]
    r = np.zeros_like(image).astype(np.uint8)
    g = np.zeros_like(image).astype(np.uint8)
    b = np.zeros_like(image).astype(np.uint8)
    r[image == 1], g[image == 1], b[image == 1] = colours[random.randrange(0, 10)]
    coloured_mask = np.stack([r, g, b], axis=2)
    return coloured_mask


def instance_segmentation_api(result, threshold=0.5, rect_th=3, text_size=3, text_th=3):
    masks, boxes, pred_cls = result
    img = cv2.imread(img_path)
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    for i in range(len(masks)):
        # color mask
        rgb_mask = random_colour_masks(masks[i])
        img = cv2.addWeighted(img, 1, rgb_mask, 0.5, 0)
        # draw box
        cv2.rectangle(img, boxes[i][0], boxes[i][1], color=(0, 255, 0), thickness=rect_th)
        # class label
        cv2.putText(img, pred_cls[i], boxes[i][0],
                    cv2.FONT_HERSHEY_SIMPLEX, text_size, (0, 255, 0), thickness=text_th)


if __name__ == "__main__":
    parser = OptionParser()
    parser.add_option("--image", type=str, dest="img_path",
                      help="The path where the images are stored, "
                           "can be either a folder or an image path")
    parser.add_option("--model", type=str, dest="model_path",
                      help="Path to the TensorFlow model file")

    (options, args) = parser.parse_args(sys.argv)

    sc = init_nncontext("OpenVINO Mask_RCNN Inference Example")
    images = ImageSet.read(options.img_path, sc,
                           resize_height=800, resize_width=1365).get_image().collect()
    input_data = np.concatenate([image.reshape((1, 1) + image.shape) for image in images], axis=0)
    model_path = options.model_path
    model = InferenceModel()
    model.load_openvino(model_path,
                        weight_path=model_path[:model_path.rindex(".")] +
                        ".bin")
    predictions = model.predict(input_data)
    # Print the detection result of the first image.
    print(predictions[0])
    print("finished...")
    sc.stop()
